src/Stratis.Bitcoin.Features.Consensus/CoinViews/CachedCoinView.cs

```diff
@@ -398,19 +416,24 @@ public async Task<uint256> Rewind()
 
             using (await this.lockobj.LockAsync().ConfigureAwait(false))
             {
-                if (this.blockHash == this.innerBlockHash)
-                    this.unspents.Clear();
-
-                if (this.unspents.Count != 0)
+                // Check if rewind data is available in local cache. If it is 
+                // we can rewind and there is no need to check underlying storage.
+                if (this.cachedRewindDataList.Count > 0)
                 {
-                    // More intelligent version can restore without throwing away the cache. (as the rewind data is in the cache).
-                    this.unspents.Clear();
-                    this.blockHash = this.innerBlockHash;
+                    RewindData lastRewindData = this.cachedRewindDataList.Last();
 
-                    this.logger.LogTrace("(-)[REWOUND_TO_INNER]:'{0}'", this.blockHash);
+                    this.RemoveTransactions(lastRewindData);
+                    this.RestoreOutputs(lastRewindData);
+
+                    // Change current block hash to the one from the rewind data.
+                    this.blockHash = lastRewindData.PreviousBlockHash;
+
+                    this.cachedRewindDataList.RemoveAt(this.cachedRewindDataList.Count - 1);
+                    this.logger.LogTrace("(-)[REMOVED_FROM_BATCH]:'{0}'", lastRewindData.PreviousBlockHash);
```

Comment:
The argument here should match returned object, just cosmetics

Reply:
Unless there could be a race condition and it's possible that there is an another thread that might touch `this.blockHash`